Data_Engine/Create/NTree.cs

```diff
@@ -0,0 +1,179 @@
+        /***************************************************/
+        
+        public static NTree<T> NTree<T>(IEnumerable<NBound> boundingBoxes, IEnumerable<T> data, int splits = 16)
+        {
+            if (boundingBoxes.Count() != data.Count())
+                return null;
+
+            List<Leaf<T>> leaves = boundingBoxes.Zip(data, (x, y) => new Leaf<T>() { Bounds = x, Item = y }).ToList();
+
+            return NTree(leaves, splits);
+        }
+
+        /***************************************************/
+
+        public static NTree<T> NTree<T>(IEnumerable<Leaf<T>> leaves, int splits = 16)
+        {
+            splits = splits < 2 ? 2 : splits;
+
+            if (leaves.Count() > splits)
+            {
+                int temp = (int)Math.Ceiling((decimal)(leaves.Count() / splits)) + 1;
+                temp = temp < splits ? temp : splits;
+
+
+                List<NTree<T>> safetyBranches = new List<NTree<T>>();
+
+                List<List<Leaf<T>>> subLists = new List<List<Leaf<T>>>() { leaves.ToList() };
+                while (subLists.Count < temp)
+                {
+                    // Find the one with the most items
+                    int index = LargestList(subLists);
+
+                    Output<List<Leaf<T>>, List<Leaf<T>>> split = SplitList(subLists[index]);
+
+                    subLists.RemoveAt(index);
+                    // One empty and one full list implies a singular BoundingBox for the data, which can't be stored spatially and is hence delegated to a single branch
+                    if (split.Item1.Count < 1)
+                    {
+                        safetyBranches.Add(new NTree<T>() { Items = split.Item2.ToArray(), Bounds = Bounds(split.Item2) });
+                        temp -= 2;
+                        continue;
+                    }
+                    else if (split.Item2.Count < 1)
+                    {
+                        safetyBranches.Add(new NTree<T>() { Items = split.Item1.ToArray(), Bounds = Bounds(split.Item1) });
+                        temp -= 2;
+                        continue;
+                    }
+
+                    subLists.Add(split.Item1);
+                    subLists.Add(split.Item2);
+
+                }
+
+                // Recursion
+                NTree<T>[] branches = subLists.Select(x => NTree(x, splits)).ToArray().Concat(safetyBranches).ToArray();
+
+                return new NTree<T>() { Items = branches, Bounds = Bounds(branches) };
+            }
+            else
+            {
+                return new NTree<T>() { Items = leaves.ToArray(), Bounds = Bounds(leaves) };
+            }
+        }
+
+
+        /***************************************************/
+        /**** Private Methods                           ****/
+        /***************************************************/
+        
+        private static Output<List<Leaf<T>>, List<Leaf<T>>> SplitList<T>(List<Leaf<T>> list, int sampleSize = 60)
```

Comment:
`sampleSize` = amount of object to use for finding the centre for splitting into different boxes.

It's both not exposed for the user atm.

What we'd really want would be the median point of all the `NBounds`, which would be very computational expensive. An OK approximation is the average of a sample of the boxes, which is what I'm taking here.